File: AllSamples/Samples/Samples_Source/peer networking/pnrp resolver/vb/pnrpresolver.vb
'=====================================================================
' File:      PnrpResolver.vb
'
' Summary:   Sample application to resolve a PeerName.
'
'---------------------------------------------------------------------
' This file is part of the Microsoft .NET Framework SDK Code Samples.
' 
' Copyright (C) Microsoft Corporation.  All rights reserved.
' 
' This source code is intended only as a supplement to Microsoft
' Development Tools and/or on-line documentation.  See these other
' materials for detailed information regarding Microsoft code samples.
' 
' THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
' KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
' IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
' PARTICULAR PURPOSE.
'---------------------------------------------------------------------

Option Explicit On
Option Strict On

Imports System.Collections.Generic
Imports System.Net
Imports System.Net.PeerToPeer
Imports System.Windows.Forms

<DefaultInstanceProperty("GetInstance")> Public Class FormMain
    Inherits System.Windows.Forms.Form

    Private resolver As PnrpEndPointResolver

    Private panel As System.Windows.Forms.Panel
    Private labelPeerName As System.Windows.Forms.Label
    Private textBoxPeerName As System.Windows.Forms.TextBox
    Private labelCloud As System.Windows.Forms.Label
    Private dropListCloud As System.Windows.Forms.ComboBox
    Private labelMaxTime As System.Windows.Forms.Label
    Private textBoxTimeout As System.Windows.Forms.NumericUpDown
    Private labelMaxResults As System.Windows.Forms.Label
    Private textBoxMaxResults As System.Windows.Forms.NumericUpDown
    Private labelCriteria As System.Windows.Forms.Label
    Private dropListCriteria As System.Windows.Forms.ComboBox
    Private WithEvents buttonResolve As System.Windows.Forms.Button
    Private textBoxResults As System.Windows.Forms.TextBox

    Private components As System.ComponentModel.Container = Nothing


    ' Constructor for this form.
    '
    Public Sub New()
        InitializeComponent()

        ' Initialize the lists.
        InitializeCloudList()
        InitializeCriteria()
        InitializeResolver()

    End Sub 'New

    ' Clean up any resources being used.
    '
    Protected Overrides Sub Dispose(ByVal disposing As Boolean)
        If disposing Then
            If Not (components Is Nothing) Then
                components.Dispose()
            End If
        End If
        MyBase.Dispose(disposing)

    End Sub 'Dispose

#Region "Windows Form Designer generated code"

    ' <summary>
    ' Required method for Designer support - do not modify
    ' the contents of this method with the code editor.
    ' </summary>
    Private Sub InitializeComponent()
        Me.panel = New System.Windows.Forms.Panel
        Me.dropListCriteria = New System.Windows.Forms.ComboBox
        Me.labelCriteria = New System.Windows.Forms.Label
        Me.labelMaxTime = New System.Windows.Forms.Label
        Me.dropListCloud = New System.Windows.Forms.ComboBox
        Me.labelPeerName = New System.Windows.Forms.Label
        Me.textBoxPeerName = New System.Windows.Forms.TextBox
        Me.labelMaxResults = New System.Windows.Forms.Label
        Me.textBoxMaxResults = New System.Windows.Forms.NumericUpDown
        Me.textBoxTimeout = New System.Windows.Forms.NumericUpDown
        Me.labelCloud = New System.Windows.Forms.Label
        Me.buttonResolve = New System.Windows.Forms.Button
        Me.textBoxResults = New System.Windows.Forms.TextBox
        Me.panel.SuspendLayout()
        CType(Me.textBoxMaxResults, System.ComponentModel.ISupportInitialize).BeginInit()
        CType(Me.textBoxTimeout, System.ComponentModel.ISupportInitialize).BeginInit()
        Me.SuspendLayout()
        '
        'panel
        '
        Me.panel.Controls.Add(Me.dropListCriteria)
        Me.panel.Controls.Add(Me.labelCriteria)
        Me.panel.Controls.Add(Me.labelMaxTime)
        Me.panel.Controls.Add(Me.dropListCloud)
        Me.panel.Controls.Add(Me.labelPeerName)
        Me.panel.Controls.Add(Me.textBoxPeerName)
        Me.panel.Controls.Add(Me.labelMaxResults)
        Me.panel.Controls.Add(Me.textBoxMaxResults)
        Me.panel.Controls.Add(Me.textBoxTimeout)
        Me.panel.Controls.Add(Me.labelCloud)
        Me.panel.Controls.Add(Me.buttonResolve)
        Me.panel.Dock = System.Windows.Forms.DockStyle.Top
        Me.panel.Location = New System.Drawing.Point(0, 0)
        Me.panel.Name = "panel"
        Me.panel.Size = New System.Drawing.Size(517, 111)
        Me.panel.TabIndex = 0
        '
        'dropListCriteria
        '
        Me.dropListCriteria.Anchor = CType(((System.Windows.Forms.AnchorStyles.Top Or System.Windows.Forms.AnchorStyles.Left) _
                    Or System.Windows.Forms.AnchorStyles.Right), System.Windows.Forms.AnchorStyles)
        Me.dropListCriteria.DropDownStyle = System.Windows.Forms.ComboBoxStyle.DropDownList
        Me.dropListCriteria.Location = New System.Drawing.Point(76, 77)
        Me.dropListCriteria.Name = "dropListCriteria"
        Me.dropListCriteria.Size = New System.Drawing.Size(243, 21)
        Me.dropListCriteria.Sorted = True
        Me.dropListCriteria.TabIndex = 5
        '
        'labelCriteria
        '
        Me.labelCriteria.Location = New System.Drawing.Point(10, 77)
        Me.labelCriteria.Name = "labelCriteria"
        Me.labelCriteria.Size = New System.Drawing.Size(54, 16)
        Me.labelCriteria.TabIndex = 4
        Me.labelCriteria.Text = "Cr&iteria"
        '
        'labelMaxTime
        '
        Me.labelMaxTime.Anchor = CType((System.Windows.Forms.AnchorStyles.Top Or System.Windows.Forms.AnchorStyles.Right), System.Windows.Forms.AnchorStyles)
        Me.labelMaxTime.Location = New System.Drawing.Point(329, 46)
        Me.labelMaxTime.Name = "labelMaxTime"
        Me.labelMaxTime.Size = New System.Drawing.Size(94, 16)
        Me.labelMaxTime.TabIndex = 8
        Me.labelMaxTime.Text = "Maximum &Time:"
        '
        'dropListCloud
        '
        Me.dropListCloud.Anchor = CType(((System.Windows.Forms.AnchorStyles.Top Or System.Windows.Forms.AnchorStyles.Left) _
                    Or System.Windows.Forms.AnchorStyles.Right), System.Windows.Forms.AnchorStyles)
        Me.dropListCloud.DropDownStyle = System.Windows.Forms.ComboBoxStyle.DropDownList
        Me.dropListCloud.Location = New System.Drawing.Point(76, 46)
        Me.dropListCloud.Name = "dropListCloud"
        Me.dropListCloud.Size = New System.Drawing.Size(243, 21)
        Me.dropListCloud.Sorted = True
        Me.dropListCloud.TabIndex = 3
        '
        'labelPeerName
        '
        Me.labelPeerName.Location = New System.Drawing.Point(10, 20)
        Me.labelPeerName.Name = "labelPeerName"
        Me.labelPeerName.Size = New System.Drawing.Size(61, 16)
        Me.labelPeerName.TabIndex = 0
        Me.labelPeerName.Text = "&PeerName:"
        '
        'textBoxPeerName
        '
        Me.textBoxPeerName.Anchor = CType(((System.Windows.Forms.AnchorStyles.Top Or System.Windows.Forms.AnchorStyles.Left) _
                    Or System.Windows.Forms.AnchorStyles.Right), System.Windows.Forms.AnchorStyles)
        Me.textBoxPeerName.Location = New System.Drawing.Point(76, 17)
        Me.textBoxPeerName.MaxLength = 255
        Me.textBoxPeerName.Name = "textBoxPeerName"
        Me.textBoxPeerName.Size = New System.Drawing.Size(243, 19)
        Me.textBoxPeerName.TabIndex = 1
        Me.textBoxPeerName.Text = "0.0"
        '
        'labelMaxResults
        '
        Me.labelMaxResults.Anchor = CType((System.Windows.Forms.AnchorStyles.Top Or System.Windows.Forms.AnchorStyles.Right), System.Windows.Forms.AnchorStyles)
        Me.labelMaxResults.Location = New System.Drawing.Point(329, 16)
        Me.labelMaxResults.Name = "labelMaxResults"
        Me.labelMaxResults.Size = New System.Drawing.Size(98, 16)
        Me.labelMaxResults.TabIndex = 6
        Me.labelMaxResults.Text = "Maximum &Results:"
        '
        'textBoxMaxResults
        '
        Me.textBoxMaxResults.Anchor = CType((System.Windows.Forms.AnchorStyles.Top Or System.Windows.Forms.AnchorStyles.Right), System.Windows.Forms.AnchorStyles)
        Me.textBoxMaxResults.Location = New System.Drawing.Point(429, 16)
        Me.textBoxMaxResults.Maximum = New Decimal(New Integer() {10000, 0, 0, 0})
        Me.textBoxMaxResults.Minimum = New Decimal(New Integer() {1, 0, 0, 0})
        Me.textBoxMaxResults.Name = "textBoxMaxResults"
        Me.textBoxMaxResults.Size = New System.Drawing.Size(75, 20)
        Me.textBoxMaxResults.TabIndex = 7
        Me.textBoxMaxResults.Value = New Decimal(New Integer() {999, 0, 0, 0})
        '
        'textBoxTimeout
        '
        Me.textBoxTimeout.Anchor = CType((System.Windows.Forms.AnchorStyles.Top Or System.Windows.Forms.AnchorStyles.Right), System.Windows.Forms.AnchorStyles)
        Me.textBoxTimeout.Location = New System.Drawing.Point(429, 46)
        Me.textBoxTimeout.Maximum = New Decimal(New Integer() {600, 0, 0, 0})
        Me.textBoxTimeout.Name = "textBoxTimeout"
        Me.textBoxTimeout.Size = New System.Drawing.Size(75, 20)
        Me.textBoxTimeout.TabIndex = 9
        Me.textBoxTimeout.Value = New Decimal(New Integer() {30, 0, 0, 0})
        '
        'labelCloud
        '
        Me.labelCloud.Location = New System.Drawing.Point(10, 46)
        Me.labelCloud.Name = "labelCloud"
        Me.labelCloud.Size = New System.Drawing.Size(38, 16)
        Me.labelCloud.TabIndex = 2
        Me.labelCloud.Text = "&Cloud:"
        '
        'buttonResolve
        '
        Me.buttonResolve.Anchor = CType((System.Windows.Forms.AnchorStyles.Top Or System.Windows.Forms.AnchorStyles.Right), System.Windows.Forms.AnchorStyles)
        Me.buttonResolve.Location = New System.Drawing.Point(429, 77)
        Me.buttonResolve.Name = "buttonResolve"
        Me.buttonResolve.TabIndex = 10
        Me.buttonResolve.Text = "Resolve"
        '
        'textBoxResults
        '
        Me.textBoxResults.Dock = System.Windows.Forms.DockStyle.Fill
        Me.textBoxResults.Location = New System.Drawing.Point(0, 111)
        Me.textBoxResults.Multiline = True
        Me.textBoxResults.Name = "textBoxResults"
        Me.textBoxResults.ScrollBars = System.Windows.Forms.ScrollBars.Vertical
        Me.textBoxResults.Size = New System.Drawing.Size(517, 200)
        Me.textBoxResults.TabIndex = 1
        Me.textBoxResults.TabStop = False
        '
        'FormMain
        '
        Me.AcceptButton = Me.buttonResolve
        Me.AutoScaleBaseSize = New System.Drawing.Size(5, 13)
        Me.ClientSize = New System.Drawing.Size(517, 311)
        Me.Controls.Add(Me.textBoxResults)
        Me.Controls.Add(Me.panel)
        Me.MaximizeBox = False
        Me.MinimumSize = New System.Drawing.Size(300, 200)
        Me.Name = "FormMain"
        Me.Text = "PNRP Resolver"
        Me.panel.ResumeLayout(False)
        Me.panel.PerformLayout()
        CType(Me.textBoxMaxResults, System.ComponentModel.ISupportInitialize).EndInit()
        CType(Me.textBoxTimeout, System.ComponentModel.ISupportInitialize).EndInit()
        Me.ResumeLayout(False)
        Me.PerformLayout()

    End Sub 'InitializeComponent

        Friend Shared ReadOnly Property GetInstance() As FormMain
            Get
                If m_DefaultInstance Is Nothing OrElse m_DefaultInstance.IsDisposed() Then
                    SyncLock m_SyncObject
                        If m_DefaultInstance Is Nothing OrElse m_DefaultInstance.IsDisposed() Then
                            m_DefaultInstance = New FormMain
                        End If
                    End SyncLock
                End If
                Return m_DefaultInstance
            End Get
        End Property

        Private Shared m_DefaultInstance As FormMain
        Private Shared m_SyncObject As New Object

#End Region


    ' The main entry point for the application.
    '
    <STAThread()> _
    Shared Sub Main()
        Application.EnableVisualStyles()
        Application.Run(New FormMain)
    End Sub 'Main


    ' Initialize the Cloud drop down list.
    '
    Private Sub InitializeCloudList()
        Dim clouds As List(Of Cloud)
        clouds = CloudWatcher.GetClouds()

        Dim cloud As Cloud
        For Each cloud In clouds
            dropListCloud.Items.Add(cloud)
        Next cloud

        ' Always select the first one (Global)
        dropListCloud.SelectedIndex = 0

    End Sub 'InitializeCloudList

    ' Initialize the ResolutionCriteria drop down list.
    '
    Private Sub InitializeCriteria()
        Dim obj As [Object]
        For Each obj In [Enum].GetValues(GetType(ResolutionCriteria))
            dropListCriteria.Items.Add(obj)
        Next obj

        dropListCriteria.SelectedIndex = 0
    End Sub 'InitializeCriteria

    ' Initialize the global resolver object and associated callbacks.
    '
    Private Sub InitializeResolver()
        resolver = New PnrpEndPointResolver
        resolver.SynchronizingObject = Me
        AddHandler resolver.PeerNameFound, AddressOf OnPeerNameFound
        AddHandler resolver.ResolutionCompleted, AddressOf OnResolutionCompleted
    End Sub 'InitializeResolver

    ' Retrieve the currently selected cloud.
    '
    Public ReadOnly Property SelectedCloud() As Cloud
        Get
            Return CType(dropListCloud.SelectedItem, Cloud)
        End Get
    End Property

    ' Handle the button click to start or stop the resolution process.
    '
    Private Sub buttonResolve_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles buttonResolve.Click
        If resolver.Resolving Then
            StopResolve()
        Else
            StartResolve()
        End If

        UpdateControls()

    End Sub 'buttonResolve_Click

    ' Stop the resolve process.
    '
    Private Sub StopResolve()
        If resolver.Resolving Then
            resolver.EndResolve()
        End If
    End Sub 'StopResolve

    ' Start the resolve process.
    '
    Private Sub StartResolve()
        ClearMessages()

        Try
            ' Retrieve the settings and fill out the resolver properties.
            Dim peerName As New PeerName(textBoxPeerName.Text)
            resolver.PeerName = peerName
            resolver.Cloud = SelectedCloud
            resolver.MaxResults = CType(Fix(textBoxMaxResults.Value), Integer)
            resolver.TimeOut = New TimeSpan(0, 0, CType(Fix(textBoxTimeout.Value), Integer))
            resolver.ResolutionCriteria = CType(dropListCriteria.SelectedItem, ResolutionCriteria)

            DisplayMessage("Resolving for " & peerName.ToString())
            resolver.BeginResolve()
        Catch e As Exception
            ' Display all exceptions.
            MessageBox.Show(Me, e.Message & vbCrLf & vbCrLf & e.StackTrace, _
                "Exception from " & e.Source, MessageBoxButtons.OK, MessageBoxIcon.Error)
        End Try

    End Sub 'StartResolve

    ' Enable (or disable) the controls.
    '
    Private Sub UpdateControls()
        ' Disable the controls if we are resolving
        Dim resolving As Boolean = resolver.Resolving

        Dim control As Control
        For Each control In panel.Controls
            If Not TypeOf control Is Label Then
                If Not resolving Then
                    If TypeOf control.Tag Is Boolean Then
                        ' Restore the previous state
                        control.Enabled = CBool(control.Tag)
                    End If
                Else
                    ' Save the previous state
                    control.Tag = control.Enabled
                    control.Enabled = False
                End If
            End If
        Next control

        ' Update the button (always enabled)
        If (resolving) Then
            buttonResolve.Text = "Stop"
        Else
            buttonResolve.Text = "Resolve"
        End If
        buttonResolve.Enabled = True

    End Sub 'UpdateControls

    ' Handle the event of finding a result.
    '
    Private Sub OnPeerNameFound(ByVal sender As Object, ByVal e As PeerNameFoundEventArgs)
        DisplayMessage("")
        DisplayMessage("Found result with comment='" & e.PnrpEndPoint.Comment & "'")

        ' Display each address
        Dim ipEndPoints As IPEndPoint() = e.PnrpEndPoint.IPEndPoints
        Dim i As Integer
        For i = 1 To ipEndPoints.Length
            DisplayMessage("  IPEndPoint " & i.ToString() & ": " & e.PnrpEndPoint.IPEndPoints(i - 1).ToString())
        Next i

    End Sub 'OnPeerNameFound

    ' Handle the event that is raised when the resolve is complete.
    '
    Private Sub OnResolutionCompleted(ByVal sender As Object, ByVal e As ResolutionCompletedEventArgs)
        DisplayMessage("")
        DisplayMessage("Resolve Completed: " & e.Reason.ToString())

        UpdateControls()
    End Sub 'OnResolutionCompleted

    ' Display a message in the text control.
    '
    Private Sub DisplayMessage(ByVal message As String)
        textBoxResults.AppendText(message)
        textBoxResults.AppendText(vbCrLf)
    End Sub 'DisplayMessage

    ' Clear the message area.
    '
    Private Sub ClearMessages()
        textBoxResults.Clear()
    End Sub 'ClearMessages

End Class 'FormMain

File: AllSamples/Samples/Samples_Source/peer networking/pnrp resolver/cs/pnrpresolver.cs
//=====================================================================
// File:      PnrpResolver.cs
//
// Summary:   Sample application to resolve a PeerName.
//
//---------------------------------------------------------------------
// This file is part of the Microsoft .NET Framework SDK Code Samples.
// 
// Copyright (C) Microsoft Corporation.  All rights reserved.
// 
// This source code is intended only as a supplement to Microsoft
// Development Tools and/or on-line documentation.  See these other
// materials for detailed information regarding Microsoft code samples.
// 
// THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
// KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
// PARTICULAR PURPOSE.
//---------------------------------------------------------------------

using System;
using System.Drawing;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Net;
using System.Net.PeerToPeer;
using System.Windows.Forms;

namespace PnrpResolver
{
	/// <summary>
	/// Sample application to resolve for PeerNames.
	/// </summary>
	public class FormMain : System.Windows.Forms.Form
    {
        private PnrpEndPointResolver resolver;

        private System.Windows.Forms.Panel panel;
        private System.Windows.Forms.Label labelPeerName;
        private System.Windows.Forms.TextBox textBoxPeerName;
        private System.Windows.Forms.Label labelCloud;
        private System.Windows.Forms.ComboBox dropListCloud;
        private System.Windows.Forms.Label labelMaxTime;
        private System.Windows.Forms.NumericUpDown textBoxTimeout;
        private System.Windows.Forms.Label labelMaxResults;
        private System.Windows.Forms.NumericUpDown textBoxMaxResults;
        private System.Windows.Forms.Label labelCriteria;
        private System.Windows.Forms.ComboBox dropListCriteria;
        private System.Windows.Forms.Button buttonResolve;
        private System.Windows.Forms.TextBox textBoxResults;

		private System.ComponentModel.Container components = null;

        /// <summary>
        /// Constructor for this form.
        /// </summary>
		public FormMain()
        {		    
		    InitializeComponent();

            // Initialize the lists.
		    InitializeCloudList();
		    InitializeCriteria();
		    InitializeResolver();
		}

		/// <summary>
		/// Clean up any resources being used.
		/// </summary>
		protected override void Dispose( bool disposing )
		{
			if( disposing )
			{
				if (components != null)
				{
					components.Dispose();
				}
			}
			base.Dispose( disposing );
		}

        #region Windows Form Designer generated code
        /// <summary>
        /// Required method for Designer support - do not modify
        /// the contents of this method with the code editor.
        /// </summary>
        private void InitializeComponent()
        {
            this.panel = new System.Windows.Forms.Panel();
            this.dropListCriteria = new System.Windows.Forms.ComboBox();
            this.labelCriteria = new System.Windows.Forms.Label();
            this.labelMaxTime = new System.Windows.Forms.Label();
            this.dropListCloud = new System.Windows.Forms.ComboBox();
            this.labelPeerName = new System.Windows.Forms.Label();
            this.textBoxPeerName = new System.Windows.Forms.TextBox();
            this.labelMaxResults = new System.Windows.Forms.Label();
            this.textBoxMaxResults = new System.Windows.Forms.NumericUpDown();
            this.textBoxTimeout = new System.Windows.Forms.NumericUpDown();
            this.labelCloud = new System.Windows.Forms.Label();
            this.buttonResolve = new System.Windows.Forms.Button();
            this.textBoxResults = new System.Windows.Forms.TextBox();
            this.panel.SuspendLayout();
            ((System.ComponentModel.ISupportInitialize)(this.textBoxMaxResults)).BeginInit();
            ((System.ComponentModel.ISupportInitialize)(this.textBoxTimeout)).BeginInit();
            this.SuspendLayout();

            // 
            // panel
            // 
            this.panel.Controls.Add(this.dropListCriteria);
            this.panel.Controls.Add(this.labelCriteria);
            this.panel.Controls.Add(this.labelMaxTime);
            this.panel.Controls.Add(this.dropListCloud);
            this.panel.Controls.Add(this.labelPeerName);
            this.panel.Controls.Add(this.textBoxPeerName);
            this.panel.Controls.Add(this.labelMaxResults);
            this.panel.Controls.Add(this.textBoxMaxResults);
            this.panel.Controls.Add(this.textBoxTimeout);
            this.panel.Controls.Add(this.labelCloud);
            this.panel.Controls.Add(this.buttonResolve);
            this.panel.Dock = System.Windows.Forms.DockStyle.Top;
            this.panel.Location = new System.Drawing.Point(0, 0);
            this.panel.Name = "panel";
            this.panel.Size = new System.Drawing.Size(517, 111);
            this.panel.TabIndex = 0;

            // 
            // dropListCriteria
            // 
            this.dropListCriteria.Anchor = ((System.Windows.Forms.AnchorStyles)(((System.Windows.Forms.AnchorStyles.Top | System.Windows.Forms.AnchorStyles.Left) | System.Windows.Forms.AnchorStyles.Right)));
            this.dropListCriteria.DropDownStyle = System.Windows.Forms.ComboBoxStyle.DropDownList;
            this.dropListCriteria.Location = new System.Drawing.Point(76, 77);
            this.dropListCriteria.Name = "dropListCriteria";
            this.dropListCriteria.Size = new System.Drawing.Size(243, 21);
            this.dropListCriteria.Sorted = true;
            this.dropListCriteria.TabIndex = 5;

            // 
            // labelCriteria
            // 
            this.labelCriteria.Location = new System.Drawing.Point(10, 77);
            this.labelCriteria.Name = "labelCriteria";
            this.labelCriteria.Size = new System.Drawing.Size(54, 16);
            this.labelCriteria.TabIndex = 4;
            this.labelCriteria.Text = "Cr&iteria";

            // 
            // labelMaxTime
            // 
            this.labelMaxTime.Anchor = ((System.Windows.Forms.AnchorStyles)((System.Windows.Forms.AnchorStyles.Top | System.Windows.Forms.AnchorStyles.Right)));
            this.labelMaxTime.Location = new System.Drawing.Point(329, 46);
            this.labelMaxTime.Name = "labelMaxTime";
            this.labelMaxTime.Size = new System.Drawing.Size(94, 16);
            this.labelMaxTime.TabIndex = 8;
            this.labelMaxTime.Text = "Maximum &Time:";

            // 
            // dropListCloud
            // 
            this.dropListCloud.Anchor = ((System.Windows.Forms.AnchorStyles)(((System.Windows.Forms.AnchorStyles.Top | System.Windows.Forms.AnchorStyles.Left) | System.Windows.Forms.AnchorStyles.Right)));
            this.dropListCloud.DropDownStyle = System.Windows.Forms.ComboBoxStyle.DropDownList;
            this.dropListCloud.Location = new System.Drawing.Point(76, 46);
            this.dropListCloud.Name = "dropListCloud";
            this.dropListCloud.Size = new System.Drawing.Size(243, 21);
            this.dropListCloud.Sorted = true;
            this.dropListCloud.TabIndex = 3;

            // 
            // labelPeerName
            // 
            this.labelPeerName.Location = new System.Drawing.Point(10, 20);
            this.labelPeerName.Name = "labelPeerName";
            this.labelPeerName.Size = new System.Drawing.Size(61, 16);
            this.labelPeerName.TabIndex = 0;
            this.labelPeerName.Text = "&PeerName:";

            // 
            // textBoxPeerName
            // 
            this.textBoxPeerName.Anchor = ((System.Windows.Forms.AnchorStyles)(((System.Windows.Forms.AnchorStyles.Top | System.Windows.Forms.AnchorStyles.Left) | System.Windows.Forms.AnchorStyles.Right)));
            this.textBoxPeerName.Location = new System.Drawing.Point(76, 17);
            this.textBoxPeerName.MaxLength = 255;
            this.textBoxPeerName.Name = "textBoxPeerName";
            this.textBoxPeerName.Size = new System.Drawing.Size(243, 19);
            this.textBoxPeerName.TabIndex = 1;
            this.textBoxPeerName.Text = "0.0";

            // 
            // labelMaxResults
            // 
            this.labelMaxResults.Anchor = ((System.Windows.Forms.AnchorStyles)((System.Windows.Forms.AnchorStyles.Top | System.Windows.Forms.AnchorStyles.Right)));
            this.labelMaxResults.Location = new System.Drawing.Point(329, 16);
            this.labelMaxResults.Name = "labelMaxResults";
            this.labelMaxResults.Size = new System.Drawing.Size(98, 16);
            this.labelMaxResults.TabIndex = 6;
            this.labelMaxResults.Text = "Maximum &Results:";

            // 
            // textBoxMaxResults
            // 
            this.textBoxMaxResults.Anchor = ((System.Windows.Forms.AnchorStyles)((System.Windows.Forms.AnchorStyles.Top | System.Windows.Forms.AnchorStyles.Right)));
            this.textBoxMaxResults.Location = new System.Drawing.Point(429, 16);
            this.textBoxMaxResults.Maximum = new System.Decimal(new int[] {
                10000, 0, 0, 0
            });
            this.textBoxMaxResults.Minimum = new System.Decimal(new int[] {
                1, 0, 0, 0
            });
            this.textBoxMaxResults.Name = "textBoxMaxResults";
            this.textBoxMaxResults.Size = new System.Drawing.Size(75, 20);
            this.textBoxMaxResults.TabIndex = 7;
            this.textBoxMaxResults.Value = new System.Decimal(new int[] {
                999, 0, 0, 0
            });

            // 
            // textBoxTimeout
            // 
            this.textBoxTimeout.Anchor = ((System.Windows.Forms.AnchorStyles)((System.Windows.Forms.AnchorStyles.Top | System.Windows.Forms.AnchorStyles.Right)));
            this.textBoxTimeout.Location = new System.Drawing.Point(429, 46);
            this.textBoxTimeout.Maximum = new System.Decimal(new int[] {
                600, 0, 0, 0
            });
            this.textBoxTimeout.Name = "textBoxTimeout";
            this.textBoxTimeout.Size = new System.Drawing.Size(75, 20);
            this.textBoxTimeout.TabIndex = 9;
            this.textBoxTimeout.Value = new System.Decimal(new int[] {
                30, 0, 0, 0
            });

            // 
            // labelCloud
            // 
            this.labelCloud.Location = new System.Drawing.Point(10, 46);
            this.labelCloud.Name = "labelCloud";
            this.labelCloud.Size = new System.Drawing.Size(38, 16);
            this.labelCloud.TabIndex = 2;
            this.labelCloud.Text = "&Cloud:";

            // 
            // buttonResolve
            // 
            this.buttonResolve.Anchor = ((System.Windows.Forms.AnchorStyles)((System.Windows.Forms.AnchorStyles.Top | System.Windows.Forms.AnchorStyles.Right)));
            this.buttonResolve.Location = new System.Drawing.Point(429, 77);
            this.buttonResolve.Name = "buttonResolve";
            this.buttonResolve.TabIndex = 10;
            this.buttonResolve.Text = "Resolve";
            this.buttonResolve.Click += new System.EventHandler(this.buttonResolve_Click);

            // 
            // textBoxResults
            // 
            this.textBoxResults.Dock = System.Windows.Forms.DockStyle.Fill;
            this.textBoxResults.Location = new System.Drawing.Point(0, 111);
            this.textBoxResults.Multiline = true;
            this.textBoxResults.Name = "textBoxResults";
            this.textBoxResults.ScrollBars = System.Windows.Forms.ScrollBars.Vertical;
            this.textBoxResults.Size = new System.Drawing.Size(517, 200);
            this.textBoxResults.TabIndex = 1;
            this.textBoxResults.TabStop = false;

            // 
            // FormMain
            // 
            this.AcceptButton = this.buttonResolve;
            this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
            this.ClientSize = new System.Drawing.Size(517, 311);
            this.Controls.Add(this.textBoxResults);
            this.Controls.Add(this.panel);
            this.MaximizeBox = false;
            this.MinimumSize = new System.Drawing.Size(300, 200);
            this.Name = "FormMain";
            this.Text = "PNRP Resolver";
            this.panel.ResumeLayout(false);
            this.panel.PerformLayout();
            ((System.ComponentModel.ISupportInitialize)(this.textBoxMaxResults)).EndInit();
            ((System.ComponentModel.ISupportInitialize)(this.textBoxTimeout)).EndInit();
            this.ResumeLayout(false);
            this.PerformLayout();
        }

        #endregion
        
        /// <summary>
        /// The main entry point for the application.
        /// </summary>
        [STAThread]
        static void Main()
        {
            Application.EnableVisualStyles();
            Application.Run(new FormMain());
        }

        /// <summary>
        /// Initialize the Cloud drop down list.
        /// </summary>
        private void InitializeCloudList()
        {
            List<Cloud> clouds = CloudWatcher.GetClouds();

            foreach (Cloud cloud in clouds)
            {
                dropListCloud.Items.Add(cloud);
            }

            // Always select the first one (Global)
            dropListCloud.SelectedIndex = 0;
        }

        /// <summary>
        /// Initialize the ResolutionCriteria drop down list.
        /// </summary>
        private void InitializeCriteria()
        {
            foreach (Object obj in Enum.GetValues(typeof(ResolutionCriteria)))
            {
                dropListCriteria.Items.Add(obj);
            }

            dropListCriteria.SelectedIndex = 0;
        }

        /// <summary>
        /// Initialize the global resolver object and associated callbacks.
        /// </summary>
        private void InitializeResolver()
        {
            resolver = new PnrpEndPointResolver();
            resolver.SynchronizingObject = this;
            resolver.PeerNameFound += new PeerNameFoundEventHandler(OnPeerNameFound);
            resolver.ResolutionCompleted += new ResolutionCompletedEventHandler(OnResolutionCompleted);
        }


        /// <summary>
        /// Retrieve the currently selected cloud.
        /// </summary>
        /// <value></value>
        public Cloud SelectedCloud
        {
            get
            {
                return (Cloud)dropListCloud.SelectedItem;
            }
        }

        /// <summary>
        /// Handle the button click to start or stop the resolution process.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void buttonResolve_Click(object sender, System.EventArgs e)
        {
            if (resolver.Resolving)
            {
                StopResolve();
            }
            else
            {
                StartResolve();
            }

            UpdateControls();
        }

        /// <summary>
        /// Stop the resolve process.
        /// </summary>
        private void StopResolve()
        {
            if (resolver.Resolving)
            {
                resolver.EndResolve();
            }
        }

        /// <summary>
        /// Start the resolve process.
        /// </summary>
        private void StartResolve()
        {
            ClearMessages();

            try
            {
                // Retrieve the settings and fill out the resolver properties.
                PeerName peerName = new PeerName(textBoxPeerName.Text);
                resolver.PeerName = peerName;
                resolver.Cloud = SelectedCloud;
                resolver.MaxResults = (int) textBoxMaxResults.Value;
                resolver.TimeOut = new TimeSpan(0, 0, (int) textBoxTimeout.Value);
                resolver.ResolutionCriteria = (ResolutionCriteria)dropListCriteria.SelectedItem;

                DisplayMessage("Resolving for " + peerName.ToString());
                resolver.BeginResolve();
            }
            catch (Exception e)
            {
                // Display all exceptions.
                MessageBox.Show(this, e.Message + "\r\n\r\n" + e.StackTrace,
                    "Exception from " + e.Source, MessageBoxButtons.OK, MessageBoxIcon.Error);
            }
        }
        
        /// <summary>
        /// Enable (or disable) the controls.
        /// </summary>
        private void UpdateControls()
        {
            // Disable the controls if we are resolving
            bool resolving = resolver.Resolving;

            foreach (Control control in panel.Controls)
            {
                if (!(control is Label))
                {
                    if (!resolving)
                    {
                        if (control.Tag is bool)
                        {
                            // Restore the previous state
                            control.Enabled = (bool)control.Tag;
                        }
                    }
                    else
                    {
                        // Save the previous state
                        control.Tag = control.Enabled;
                        control.Enabled = false;
                    }
                }
            }

            // Update the button (always enabled)
            buttonResolve.Text = resolving ? "Stop" : "Resolve";
            buttonResolve.Enabled = true;
        }

        /// <summary>
        /// Handle the event of finding a result.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void OnPeerNameFound(object sender, PeerNameFoundEventArgs e)
        {
            DisplayMessage("");
            DisplayMessage("Found result with comment='" + e.PnrpEndPoint.Comment + "'");

            // Display each address
            IPEndPoint[] ipEndPoints = e.PnrpEndPoint.IPEndPoints;
            for (int i = 0; i < ipEndPoints.Length; i++)
            {
                DisplayMessage("  IPEndPoint " + (i+1).ToString()
                    + ": " + e.PnrpEndPoint.IPEndPoints[i].ToString());
            }
        }

        /// <summary>
        /// Handle the event that is raised when the resolve is complete.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void OnResolutionCompleted(object sender, ResolutionCompletedEventArgs e)
        {
            DisplayMessage("");
            DisplayMessage("Resolve Completed: " + e.Reason.ToString());

            UpdateControls();
        }

        /// <summary>
        /// Display a message in the text control.
        /// </summary>
        /// <param name="message">The text message to display.</param>
        private void DisplayMessage(string message)
        {
            textBoxResults.AppendText(message);
            textBoxResults.AppendText("\r\n");
        }

        /// <summary>
        /// Clear the message area.
        /// </summary>
        private void ClearMessages()
        {
            textBoxResults.Clear();
        }
    }
}
